File: Tokens/bikemarket/workflows/src/main/java/net/corda/samples/bikemarket/flows/IssueNewBike.java
package net.corda.samples.bikemarket.flows;

import co.paralleluniverse.fibers.Suspendable;
import com.google.common.collect.ImmutableList;
import com.r3.corda.lib.tokens.contracts.states.NonFungibleToken;
import com.r3.corda.lib.tokens.workflows.flows.rpc.IssueTokens;
import com.r3.corda.lib.tokens.workflows.utilities.NonFungibleTokenBuilder;
import net.corda.samples.bikemarket.states.FrameTokenState;
import net.corda.samples.bikemarket.states.WheelsTokenState;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.flows.FlowException;
import net.corda.core.flows.FlowLogic;
import net.corda.core.flows.InitiatingFlow;
import net.corda.core.flows.StartableByRPC;
import net.corda.core.identity.Party;
import net.corda.core.transactions.SignedTransaction;

@InitiatingFlow
@StartableByRPC
public class IssueNewBike extends FlowLogic<String> {

    private final String frameSerial;
    private final String wheelsSerial;
    private final Party holder;

    public IssueNewBike(String frameSerial, String wheelsSerial, Party holder) {
        this.frameSerial = frameSerial;
        this.wheelsSerial = wheelsSerial;
        this.holder = holder;
    }

    @Suspendable
    @Override
    public String call() throws FlowException {

        //Step 1: Frame Token
        //get frame states on ledger
        StateAndRef<FrameTokenState> frameStateAndRef = getServiceHub().getVaultService().
                queryBy(FrameTokenState.class).getStates().stream()
                .filter(sf->sf.getState().getData().getserialNum().equals(this.frameSerial)).findAny()
                .orElseThrow(()-> new IllegalArgumentException("wheel token with serial=" + this.frameSerial + " not found from vault"));

        //get the TokenType object
        FrameTokenState frametokentype = frameStateAndRef.getState().getData();

        //mention the current holder also
        NonFungibleToken frametoken = new NonFungibleTokenBuilder()
                .ofTokenType(frametokentype.toPointer(frametokentype.getClass()))
                .issuedBy(getOurIdentity())
                .heldBy(holder)
                .buildNonFungibleToken();

        //Step 2: Wheels Token
        StateAndRef<WheelsTokenState> wheelStateStateAndRef = getServiceHub().getVaultService().
                queryBy(WheelsTokenState.class).getStates().stream().filter(sf->sf.getState().getData().getserialNum().equals(this.wheelsSerial)).findAny()
                .orElseThrow(()-> new IllegalArgumentException("wheel token with serial=" + this.wheelsSerial + " not found from vault"));

        //get the TokenType object
        WheelsTokenState wheeltokentype = wheelStateStateAndRef.getState().getData();

        //mention the current holder also
        NonFungibleToken wheeltoken = new NonFungibleTokenBuilder()
                .ofTokenType(wheeltokentype.toPointer(wheeltokentype.getClass()))
                .issuedBy(getOurIdentity())
                .heldBy(holder)
                .buildNonFungibleToken();

        //distribute the new bike (two token to be exact)
        //call built in flow to issue non fungible tokens
        SignedTransaction stx = subFlow(new IssueTokens(ImmutableList.of(frametoken,wheeltoken)));

        return "\nA new bike is being issued to "+ this.holder.getName().getOrganisation() + " with frame serial: "
                + this.frameSerial + "; wheels serial: "+ this.wheelsSerial + "\nTransaction ID: " + stx.getId();
    }
}

File: kotlin-samples/Tokens/bikemarket/workflows/src/main/kotlin/net/corda/samples/bikemarket/flows/IssueNewBike.kt
package net.corda.samples.bikemarket.flows

import co.paralleluniverse.fibers.Suspendable
import com.r3.corda.lib.tokens.contracts.types.TokenPointer
import com.r3.corda.lib.tokens.contracts.utilities.issuedBy
import com.r3.corda.lib.tokens.workflows.flows.rpc.IssueTokens
import com.r3.corda.lib.tokens.workflows.utilities.heldBy
import net.corda.core.flows.FlowLogic
import net.corda.core.flows.StartableByRPC
import net.corda.core.identity.Party
import net.corda.core.node.services.queryBy
import net.corda.core.utilities.ProgressTracker
import net.corda.samples.bikemarket.states.FrameTokenState
import net.corda.samples.bikemarket.states.WheelsTokenState

// *********
// * Flows *
// *********
@StartableByRPC
class IssueNewBike(val frameSerial: String,
                   val wheelsSerial: String,
                   val holder: Party) : FlowLogic<String>() {
    override val progressTracker = ProgressTracker()

    @Suspendable
    override fun call():String {
        //Step 1: Frame Token
        //get frame states on ledger
        val frameStateAndRef = serviceHub.vaultService.queryBy<FrameTokenState>().states
                .filter { it.state.data.serialNum.equals(frameSerial) }[0]

        //get the TokenType object
        val frametokentype = frameStateAndRef.state.data

        //get the pointer pointer to the frame
        val frametokenPointer = frametokentype.toPointer(frametokentype.javaClass)

        //assign the issuer to the frame type who will be issuing the tokens
        val frameissuedTokenType = frametokenPointer issuedBy ourIdentity

        //mention the current holder also
        val frametoken = frameissuedTokenType heldBy holder

        //Step 2: Wheels Token
        val wheelStateAndRef = serviceHub.vaultService.queryBy<WheelsTokenState>().states
                .filter { it.state.data.serialNum.equals(wheelsSerial) }[0]

        //get the TokenType object
        val wheeltokentype: WheelsTokenState = wheelStateAndRef.state.data

        //get the pointer pointer to the wheel
        val wheeltokenPointer: TokenPointer<*> = wheeltokentype.toPointer(wheeltokentype.javaClass)

        //assign the issuer to the wheel type who will be issuing the tokens
        val wheelissuedTokenType = wheeltokenPointer issuedBy ourIdentity

        //mention the current holder also
        val wheeltoken = wheelissuedTokenType heldBy holder

        //distribute the new bike (two token to be exact)
        //call built in flow to issue non fungible tokens
        val stx = subFlow(IssueTokens(listOf(frametoken,wheeltoken)))

        return ("\nA new bike is being issued to " + holder.name.organisation + " with frame serial: "
                + this.frameSerial + "; wheels serial: " + this.wheelsSerial + "\nTransaction ID: " + stx.id)
    }
}